### DISCUSSION TITLE
How to fix layout shift

### DISCUSSION THREAD

**Discussion by tomaszferens** [adobe/react-spectrum#2765] [Q&A]
I have a layout shift whenever user visits the app for the first time:

https://user-images.githubusercontent.com/17815179/150218429-63ea4539-f4ac-4ee8-b2e2-e9c8e3ed320e.mp4

Minimal reproduction: https://stackblitz.com/edit/nextjs-a1vca2?file=pages%2Findex.js
(you can stop the container and run the production build: `npm run build && npm run start`)

This only happens in Next.js's production build tho. In development, the issue does not occur. I actually believe this is an issue with Next and not react-spectrum, but want to make sure about it.

The way Next handles styles in development is that it injects `<style>` attributes with appropriate styles in the document's `<head>`. In production, Next generates a single CSS file that is loaded using `<link>` preload and `<link>` stylesheet.

Development:
![image](https://user-images.githubusercontent.com/17815179/150219005-d4769abe-6e47-4c18-80b9-a9ba364ca718.png)

Production:
![image](https://user-images.githubusercontent.com/17815179/150218883-a391c991-fffc-483e-8c3f-b9cd205b2ff8.png)

is there any way to fix that layout shift? 🤔

**Comment by snowystinger** [comment]
I tried out the stackblitz example you put together, but I do not see the issue you are describing. I stopped the container and ran `npm run build && npm run start` as you suggested as well, still nothing.

I can't think of anything it would be on our end, though it briefly looks like it goes from medium scale to large scale back to medium scale?
You could try putting some code into the node_modules file for the Provider and see if that's what is happening. Maybe log out around here? https://github.com/adobe/react-spectrum/blob/0f76dca420accd26c2f0f1ca73bb9ee82dc67b11/packages/%40react-spectrum/provider/src/Provider.tsx#L48
Or you could try setting the scale prop on your provider? https://react-spectrum.adobe.com/react-spectrum/Provider.html#props

**Reply by tomaszferens** [reply]
> I tried out the stackblitz example you put together, but I do not see the issue you are describing. I stopped the container and ran `npm run build && npm run start` as you suggested as well, still nothing.

You can go directly to the app in stackblitz by visiting the url, then you can open devtools, set network speed to Slow 3G, and do the reload with empty cache:

https://user-images.githubusercontent.com/17815179/150306374-17ca0fc5-a98c-43fa-96ab-831474c24ce5.mp4

As you can see the issue is still present

> I can't think of anything it would be on our end, though it briefly looks like it goes from medium scale to large scale back to medium scale? You could try putting some code into the node_modules file for the Provider and see if that's what is happening. Maybe log out around here?

> https://github.com/adobe/react-spectrum/blob/0f76dca420accd26c2f0f1ca73bb9ee82dc67b11/packages/%40react-spectrum/provider/src/Provider.tsx#L48

Will check that later, thanks!

> Or you could try setting the scale prop on your provider? https://react-spectrum.adobe.com/react-spectrum/Provider.html#props

Oh good to know, will check that as well.

**Reply by tomaszferens** [reply]
@snowystinger Looks like setting scale="medium" on Provider fixed the issue:

https://user-images.githubusercontent.com/17815179/150308341-5fe52d4e-076a-492f-9ed7-fa85c235f2fa.mp4

Thanks!!

**Reply by snowystinger** [reply]
I'm glad that fixes it, however, it seems like something else is afoot.
Unless you unset that value after hydration, your app will remain in medium scale. This isn't good for small screens or where touch is the primary mode of interaction.
I'm able to reproduce by setting the network speed.
I may not have time today, but the next thing to do would be to put in the console log I mentioned to see if the scale is changing and why it's changing.

**Reply by devongovett** [reply]
I'm not sure there is much we can do about it? There is going to be a layout shift in some cases no matter what. The server cannot know which scale to use since it cannot run media queries, so always just needs to choose one. That may end up being incorrect, and the client will need to adjust it.
